presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -589,6 +563,163 @@ public Signature resolveFunction(QualifiedName name, List<TypeSignature> paramet
+        errorMessageBuilder.append("Candidates are:\n");
+        for (ApplicableFunction function : applicableFunctions) {
+            errorMessageBuilder.append("\t * ");
+            errorMessageBuilder.append(function.getBoundSignature().toString());
+            errorMessageBuilder.append("\n");
+        }
+        throw new PrestoException(AMBIGUOUS_FUNCTION_CALL, errorMessageBuilder.toString());
+    }
+
+    private List<ApplicableFunction> identifyApplicableFunctions(List<SqlFunction> candidates, List<Type> actualParameters, boolean allowCoercion)
+    {
+        ImmutableList.Builder<ApplicableFunction> applicableFunctions = ImmutableList.builder();
+        for (SqlFunction function : candidates) {
+            Signature declaredSignature = function.getSignature();
+            Optional<Signature> boundSignature = new SignatureBinder(typeManager, declaredSignature, allowCoercion)
+                    .bind(actualParameters);
+            if (boundSignature.isPresent()) {
+                applicableFunctions.add(new ApplicableFunction(declaredSignature, boundSignature.get(), typeManager));
+            }
+        }
+        return applicableFunctions.build();
+    }
+
+    private List<ApplicableFunction> selectMostSpecificFunctions(List<ApplicableFunction> applicableFunctions, List<Type> parameters)
+    {
+        List<ApplicableFunction> mostSpecificFunctions = filterOutLessSpecificFunctions(applicableFunctions);
+        if (mostSpecificFunctions.size() > 1 && someParameterIsUnknown(parameters)) {
+            // if there is more than one function, look for functions that only cast the unknown arguments
+            List<ApplicableFunction> unknownOnlyCastFunctions = getUnknownOnlyCastFunctions(applicableFunctions, parameters);
+            if (!unknownOnlyCastFunctions.isEmpty()) {
+                mostSpecificFunctions = unknownOnlyCastFunctions;
+            }
+            if (mostSpecificFunctions.size() == 1) {
+                return mostSpecificFunctions;
+            }
+            // If the return type for all the selected function is the same, and the parameters are not declared as nullable
+            // all the functions are semantically the same. We can return just any of those.
+            if (returnTypeIsTheSame(mostSpecificFunctions) && allReturnNullOnGivenInputTypes(mostSpecificFunctions, parameters)) {
+                // make it deterministic
+                ApplicableFunction selectedFunction = Ordering.usingToString()
+                        .reverse()
+                        .sortedCopy(mostSpecificFunctions)
+                        .get(0);
+                return ImmutableList.of(selectedFunction);
+            }
+        }
+        return mostSpecificFunctions;
+    }
+
+    private List<ApplicableFunction> filterOutLessSpecificFunctions(List<ApplicableFunction> applicableFunctions)
```

Comment:
This code is invalid.
eg. if there are two candidates where first isMoreSpecific than second.
In such case both function will end up being in representatives, and we only need one of them (first one).

The code may be organized in such way if you really don't like the concept of families that covers that logic.